framework/AGGRenderer/CPPAGGRenderer/CPPAGGRenderer.cpp

```diff
@@ -81,24 +89,110 @@ namespace CPPAGGRenderer{
     renderer_aa ren_aa;
     int pngBufferSize = 0;
 
+    agg::int8u*           m_pattern;
+    agg::rendering_buffer m_pattern_rbuf;
+    agg::path_storage m_ps;
+    renderer_base_pre rb_pre;
+
     Plot(){
       rb = renderer_base(pixf);
       ren_aa = renderer_aa(rb);
+      rb_pre = renderer_base_pre(pixf_pre);
     }
 
-    void draw_solid_rect(const float *x, const float *y, float r, float g, float b, float a){
+    void draw_solid_rect(const float *x, const float *y, float r, float g, float b, float a, int hatch_pattern){
 
       agg::path_storage rect_path;
       rect_path.move_to(*x, *y);
       for (int i = 1; i < 4; i++) {
         rect_path.line_to(*(x+i),*(y+i));
       }
       rect_path.close_polygon();
-      m_ras.add_path(rect_path);
-      Color c(r, g, b, a);
-      ren_aa.color(c);
-      agg::render_scanlines(m_ras, m_sl_p8, ren_aa);
+      if (hatch_pattern == 0) {
+        Color c(r, g, b, a);
+        m_ras.add_path(rect_path);
+        ren_aa.color(c);
+        agg::render_scanlines(m_ras, m_sl_p8, ren_aa);
+      }
+      else {
+        generate_pattern(r, g, b, a, hatch_pattern);
+        typedef agg::wrap_mode_repeat_auto_pow2 wrap_x_type;
+        typedef agg::wrap_mode_repeat_auto_pow2 wrap_y_type;
+        typedef agg::image_accessor_wrap<pixfmt, wrap_x_type, wrap_y_type> img_source_type;
+        typedef agg::span_pattern_rgb<img_source_type> span_gen_type;
+        agg::span_allocator<color_type> sa;
+        pixfmt          img_pixf(m_pattern_rbuf);
+        img_source_type img_src(img_pixf);
+        span_gen_type sg(img_src, 0,0);
+        sg.alpha(span_gen_type::value_type(255.0));
+
+        m_ras.add_path(rect_path);
+        agg::render_scanlines_aa(m_ras, m_sl_p8, rb_pre, sa, sg);
+      }
+
+    }
 
+    void generate_pattern(float r, float g, float b, float a, int hatch_pattern){
```

Comment:
could this hatching be implemented in Swift, in terms of other renderer primitives?

an advantage of doing that would be that if you add later renderers, you can reuse the hatching code for them.

Reply:
@marcrasi I too had thought of that. Both AGG and SVG provide ways of filling up shapes with patterns which I why I went for that. When filling patterns with shapes we will need to handle the clipping of the pattern at the shape boundaries and maybe other hurdles. In case of SVG I think using primitives for hatching will surely make viewing the image slower because of increased parsing time because of large no. of statements to draw each primitive individually. 

Not that we need to do it this way but matplotlib also doesn't support hatching on all backbends.

What do you think?

Reply:
Yeah, this makes sense!